business/openshift_oauth.go

```diff
@@ -0,0 +1,194 @@
+// more usable on our side.
+type OAuthAuthorizationServer struct {
+	AuthorizationEndpoint string `json:"authorization_endpoint"`
+}
+
+type OAuthUser struct {
+	Metadata OAuthUserMetadata `json:"metadata"`
+}
+
+type OAuthUserMetadata struct {
+	Name string `json:"name"`
+}
+
+type OAuthRoute struct {
+	Spec OAuthRouteSpec `json:"spec"`
+}
+
+type OAuthRouteSpec struct {
+	Host string             `json:"host"`
+	TLS  *OAuthRouteTLSSpec `json:"tls,omitempty"`
+}
+
+type OAuthRouteTLSSpec struct {
+	Termination string `json:"termination"`
+}
+
+const serverPrefix = "https://openshift.default.svc/"
+
+func (in *OpenshiftOAuthService) Metadata() (metadata *OAuthMetadata, err error) {
+	var server *OAuthAuthorizationServer
+
+	if err != nil {
+		message := fmt.Errorf("Could not parse data from the Openshift API: %v", err)
+		return nil, message
+	}
+
+	response, err := request("GET", ".well-known/oauth-authorization-server", nil)
+
+	if err != nil {
+		message := fmt.Errorf("Could not parse data from the Openshift API: %v", err)
+		return nil, message
+	}
+
+	err = json.Unmarshal(response, &server)
+
+	if err != nil {
+		message := fmt.Errorf("Could not parse data from the Openshift API: %v", err)
+		return nil, message
+	}
+
+	redirectURL, err := getKialiRoutePath()
+
+	metadata = &OAuthMetadata{}
+
+	metadata.AuthorizationEndpoint = fmt.Sprintf("%s?client_id=%s&redirect_uri=%s&response_type=%s", server.AuthorizationEndpoint, "kiali", url.QueryEscape(*redirectURL), "token")
+
+	return metadata, nil
+}
+
+func (in *OpenshiftOAuthService) ValidateToken(token string) error {
+	k8sConfig, err := kubernetes.ConfigClient()
+
+	if err != nil {
+		return fmt.Errorf("could not connect to Openshift: %v", err)
+	}
+
+	k8sConfig.BearerToken = token
+
+	k8s, err := kube.NewForConfig(k8sConfig)
+
+	if err != nil {
+		return fmt.Errorf("could not connect to Openshift: %v", err)
+	}
+
+	_, err = k8s.Discovery().ServerVersion()
+
+	if err != nil {
+		return fmt.Errorf("could not get info from Openshift: %v", err)
+	}
+
+	return nil
+}
+
+func (in *OpenshiftOAuthService) GetUserInfo(token string) (*OAuthUser, error) {
+	var user *OAuthUser
+
+	response, err := request("GET", "oapi/v1/users/~", &token)
+
+	if err != nil {
+		return nil, fmt.Errorf("could not connect to Openshift: %v", err)
+	}
+
+	err = json.Unmarshal(response, &user)
+
+	if err != nil {
+		return nil, fmt.Errorf("could not connect to Openshift: %v", err)
+	}
+
+	return user, nil
+}
+
+func getKialiRoutePath() (*string, error) {
+	var route *OAuthRoute
+	var protocol string
+
+	namespace := config.Get().IstioNamespace
+
+	conf, err := kubernetes.ConfigClient()
+
+	if err != nil {
+		return nil, fmt.Errorf("could not connect to Openshift: %v", err)
+	}
+
+	response, err := request("GET", fmt.Sprintf("oapi/v1/namespaces/%s/routes/kiali", namespace), &conf.BearerToken)
+
+	err = json.Unmarshal(response, &route)
+
+	if err != nil {
+		return nil, fmt.Errorf("could not connect to Openshift: %v", err)
```

Comment:
again, bad error message. This isn't dealing with connecting to openshift is it? Looks like its parsing results - error message should say that.